src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs

```diff
@@ -962,6 +987,24 @@ private async Task<WorkUnitResult> ExecuteInstantiatedTask(ITaskExecutionHost ta
             return result;
         }
 
+        private string[] GetUndeclaredProjects(MSBuild msbuildTask)
+        {
+            if (!_componentHost.BuildParameters.IsolateProjects)
+            {
+                return Array.Empty<string>();
+            }
+
+            var declaredProjects = _taskExecutionHost.ProjectInstance.GetItems(MSBuildConstants.ProjectReferenceItemName)
+                .Select(p => FileUtilities.NormalizePath(p.EvaluatedInclude))
+                .ToHashSet();
+            declaredProjects.Add(_taskExecutionHost.ProjectInstance.FullPath);
+
+            return msbuildTask.Projects
+                .Where(p => !declaredProjects.Contains(FileUtilities.NormalizePath(p.ItemSpec)))
+                .Select(p => p.ItemSpec)
```

Comment:
This seems like it might be expensive. Do you have any perf indications either way?

Reply:
No perf experimentation, but I'll delinqify it to offset the normalization costs. In non isolated builds it gets skipped. In isolated builds, I don't expect this to be on the hot path since if there is a big p2p graph where this would method would show up (thousands of nodes), then I'd expect the parallelism to hide it.